### DISCUSSION TITLE
Gathering more scenarios

### DISCUSSION THREAD

**Discussion by pradyunsg** [pradyunsg/pip-resolver-benchmarks#8] [General]
Making a discussion topic to gather examples for pip's dependency resolution algorithm, both pathological and realistic examples.

**Comment by pradyunsg** [comment]
@notatallshaw If you find some time, I'd appreciate if you could share the requirements (+ platforms?) for a bunch of pathological cases with pip's dependency resolver!

**Reply by notatallshaw** [reply]
Yeah, I've been trying to build up some scenarios but I haven't been successful yet (I filed a seperate discussion on one of the issues).

An easy one would be `airflow-airflow[all]=={basically any version}`

Another good one: `dandi selenium`, especially if you could fetch the requirements from "2023-09-10" using `pypi-timemachine`.

I'll try and think of others that I know to be problematic from Python 3.10+.

**Reply by pradyunsg** [reply]
Scenarios with older Python versions are also good -- we can't generate their JSON files until #5 is resolved but it's probably still worth listing them here.

**Reply by notatallshaw** [reply]
Ah well, this one on Python 3.8, and doesn't even require sdists, is so highly pathological it doesn't even resolve on Pip at the moment due to a ResolutionTooDeep  error: `"numpy==1.21.6" "cython==0.29.28" "scipy>=1.4.0" "torch>=1.7" "torchaudio" "soundfile" "librosa==0.10.0.*" "numba==0.55.1" "inflect==5.6.0" "tqdm" "anyascii" "pyyaml" "fsspec>=2021.04.0" "aiohttp" "packaging" "flask" "pysbd" "pandas" "matplotlib" "trainer==0.0.20" "coqpit>=0.0.16" "pypinyin" "mecab-python3==1.0.5" "jamo" "bangla==0.0.2" "k_diffusion" "einops" "transformers"`

I have an experiemtnal branch which allows it to be resolved https://github.com/notatallshaw/pip/blob/prefer-conflicting-causes

**Reply by notatallshaw** [reply]
Here are two that fail on the current Pip because of an issue with backjumping (I have 2 alternative PRs for this on resolvelib, one that removes backjumping and one that falls back to backtracking under certain circumstances):

1.  `kedro[test]==0.18.13`
2. `"boto3>=1.28.64,<1.28.66" "dagster==1.5.5" "pandas[aws]==2.1.2" "dagster-aws>=0.21.5" "PyYAML>=6.0.1"`

Number 2 is more pathalogical than number 1 because if you remove backjumping it produces a ResolutionTooDeep error.

**Reply by notatallshaw** [reply]
Another pathalogical case is to resolve all of home assistants requirements: https://github.com/home-assistant/core/blob/dev/requirements_all.txt

**Reply by notatallshaw** [reply]
As discussed in https://github.com/pypa/pip/issues/12430, another pathological case for the current Pip resolver which I think results in "ResolutionTooDeep" (not yet let this simplified version finish on my computer):

```
sphinx
sphinx-rtd-theme
sphinx-toolbox
myst-parser
sphinxcontrib-bibtex
nbsphinx
```

This revealed a couple of issues with my experimental "prefer-conflicting-causes" branch, which has given me the opportunity to improve it which I have now done and it can quickly solve this set of requirements:  https://github.com/notatallshaw/pip/blob/prefer-conflicting-causes
